### PR TITLE
add abort step and refactored execute definition to be the same for all flow steps

### LINKED ISSUE #591: Add a step to abort the dataflow completely

### PR DESCRIPTION
**Pull request by keevan**
Resolves #591

Should improve and help with the effort in #473

### CODE REVIEW

File: classes/local/execution/engine.php [hunk 1]
```diff
@@ -470,9 +474,9 @@ public function abort(?\Throwable $reason = null) {
         $this->release_lock();
 
         // TODO: We may want to make this the responsibility of the caller.
-        if (isset($reason)) {
-            throw $reason;
-        }
+        // if (isset($reason)) {
```
Comment (Peterburnett) [inline classes/local/execution/engine.php:477]: Should we just remove this entirely? It looks like this HAS moved to the caller responsibility

File: classes/local/step/abort_trait.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        // If the condition was set, print out the results of the expression so it's obvious what it had evaluated to.
+        if ($config->condition !== '') {
+            $rawconfig = $this->get_raw_config();
+            $strresult = var_export($config->condition, true);
+            $this->log("Aborting dataflow due to the expression returning '{$strresult}': {$rawconfig->condition}");
+        }
+
+        // By default, it should abort the step.
+        throw new \Exception('Aborting');
```
Comment (jaypha) [inline classes/local/step/abort_trait.php:57]: You should merge in my abort fixes (which includes an abort step), and then change this to call engine::abort().

File: classes/local/execution/engine.php (hunk 1 above)
Comment (keevan) [reply classes/local/execution/engine.php]: I've removed this particular change, pending testing after everything has merged in and the changes might not be required anymore after Jason's other PR

File: classes/local/step/abort_trait.php (hunk 2 above)
Comment (keevan) [reply classes/local/step/abort_trait.php]: Your PR has merged in and this one has been rebased.. Thanks again.

File: classes/local/execution/engine.php [hunk 3]
```diff
@@ -631,7 +635,7 @@ public function set_status(int $status) {
 
         if ($status === self::STATUS_INITIALISED) {
             $this->log('status: ' . self::STATUS_LABELS[$status] . ', config: ' . json_encode(['isdryrun' => $this->isdryrun]));
-        } else if ($status === self::STATUS_FINALISED) {
+        } else if (in_array($status, [self::STATUS_FINALISED, self::STATUS_ABORTED], true)) {
```
Comment (jaypha) [inline classes/local/execution/engine.php:638]: Perhaps use engine::STATUS_TERMINATORS

File: lang/en/tool_dataflows.php [hunk 4]
```diff
@@ -120,6 +121,7 @@
 $string['step_name_flow_email'] = 'Flow email notification';
 $string['step_name_flow_logic_switch'] = 'Switch';
 $string['step_name_flow_logic_join'] = 'Join';
+$string['step_name_flow_abort'] = 'Abort';
```
Comment (jaypha) [inline lang/en/tool_dataflows.php:124]: Duplicate. Already defined.

File: classes/local/execution/engine.php (hunk 3 above)
Comment (keevan) [reply classes/local/execution/engine.php:638]: Thanks for the suggestion. I will take a look.

**Pull request merged**